### DISCUSSION TITLE
Unterschiede bei der Integration von gleichwertigen HM/HmIP Geräten

### DISCUSSION THREAD

**Discussion by Baxxy13** [SukramJ/aiohomematic#256] [Q&A]
Mir ist aufgefallen das gleichwertige HM/HmIP Geräte unterschiedlich dargestellt werden.
2 Beispiele:

- HmIP-SWDO (Homematic IP Fenster- und Türkontakt) & HM-Sec-SCo (Homematic Fenster- und Türkontakt, optisch) 
- HmIP-PSM (Homematic IP Smart Home Schalt-Mess-Steckdose) & HM-ES-PMSw1-Pl (Homematic Funk-Schaltaktor 1fach mit Leistungsmessung, Zwischenstecker)

Die Aktoren / Sensoren sind grundsätzlich gleichwertig in der Funktion (abgesehen von der virtuellen Kanälen der PSM, die hier aber keine Rolle spielen)

Ansicht:
![HA_Ansicht_diff_HM_HMIP](https://user-images.githubusercontent.com/54898249/152010150-05d3d779-50a1-422d-8a1e-c41b65797c57.JPG)

Die Frage die ich mir als "Neueinsteiger" stelle... Warum ist das so inkonsistent?

Die Symbole zu ändern ist ja noch einfach und schnell gemacht wenn man nicht gerade 30 HM-Sec-SCo's hat.
Aber diese 30 HM-Sec-SCo's zu "echten" Fenstersensoren zu machen erfordert schon ne Menge Fleiß und Einarbeitung in ["Customizing entities"](https://www.home-assistant.io/docs/configuration/customizing-devices/) Denn ein geschlossenes Fenster ist selten "aus".  

Wünschenswert:
Bei der Integration sollten gleichwertige HM/HmIP Geräte "out-of-the-box" auch gleich dargestellt werden. Also gleiche Symbole sowie gleicht Stati.

**Comment by SukramJ** [comment]
Da ich mir die Mühe für Geräte, die ich selber nicht besitze nicht machen wollte, gibt es gerade im klassischen HM Bereich noch Nachholbedarf.
Ich behaupte mal es geht eigentlich immer um den Parameter `STATE`, denn der bedeute je nach Modell mal dies und mal jenes. 

Wenn Du oder jemand anders sich also die Mühe macht und die semantische Verknüpfung herstellt, kann man die Inkonsistenzen natürlich einfach beheben. Aber an dem Punkt bitte ich klar um Zulieferung.

Ich nehme an die sind noch mehr Unterschiede aufgefallen?
Wenn Du mal unabhängig von deinen Gerätschaften auch leicht über den Tellrand schauen könntest, und mir eine Liste von HM Fenstersensoren und HM Autoren erstellen könntest wäre das super.

**Comment by Baxxy13** [comment]
Ok, das sollte machbar sein (wenn auch andere mithelfen).
Was brauchst du denn genau? `HssType` oder `Label` ?
Da ich auch nicht jedes Gerät habe könnte ich aber die Informationen aus den Dokumentationen zusammentragen, bzw. auch im Homematic-Forum um Hilfe bitten.

**Reply by SukramJ** [reply]
Vielleicht helfen die Listen aus pyhomematic. 
Sowas wie `HM-Sec-SC` ist notwendig. Das nennt sich in der API aber device_type.

Sensoren:
https://github.com/danielperna84/pyhomematic/blob/899a28c6036ea799cfb5513a65e37135556c60a7/pyhomematic/devicetypes/sensors.py#L1230-L1347
Actoren:
https://github.com/danielperna84/pyhomematic/blob/899a28c6036ea799cfb5513a65e37135556c60a7/pyhomematic/devicetypes/actors.py#L1081-L1252

Gepflegt wir das Ganze in [custom_homematic](https://github.com/danielperna84/custom_homematic/blob/devel/custom_components/hahm/entity_helpers.py).
Wobei ich das aber gerade überarbeite, um dort besser Listen einbinden zu können.


Ich denke das in erster Linie der Parameter STATE wichtig ist.

Bei den Sensoren denke ich in erster Linie an die Fenstersensoren (nur HM). Die meisten anderen Sensoren
haben eigenständige Parameter wie MOTION, PRESENCE, RAINING.

Bei den Aktoren sind es auch nur die Switch die STATE als Parameter haben.
Hier sollte eigentlich ein DEFAULT greifen. Das werde ich prüfen.

**Comment by Baxxy13** [comment]
Danke, das verschafft mir schon mal einen guten Überblick. Denke die Lücken lassen sich nach und nach füllen.

**Comment by Baxxy13** [comment]
Die "Sonderbehandlung" für HmIP-PS/PSM hast du ja mit der 0.29.1 rausgenommen, somit werden alle Schaltaktoren (HM/HmIP) gleich dargestellt. Gut!

### Zu den HM-Fenstersensoren.
Normal AUF/ZU (SHUTTER_CONTACT):
```
"HM-Sec-SC": ShutterContact,
"HM-Sec-SC-2": ShutterContact,
"HM-Sec-SCo": ShutterContact,
```
Drehgriffsensor (ROTARY_HANDLE_SENSOR):
```
"HM-Sec-RHS": RotaryHandleSensor,
"HM-Sec-RHS-2": RotaryHandleSensor,
"HM-Sec-xx": RotaryHandleSensor,
```
Ob der "HM-Sec-xx" tatsächlich existiert weiß ich nicht, wird aber zumindest in der HM-Datenpunkt-Doku aufgeführt.

**Reply by SukramJ** [reply]
Vielen Dank, hab ich mit 0.29.2 veröffentlicht.
Ein HM-Sec-xx gibt es in pydevccu wirklich.

**Comment by Baxxy13** [comment]
Sieht gut aus!
Dann können wir ja gleich noch über die `HmIP-SCI` (Homematic IP Smart Home Kontakt-Schnittstelle) reden.
Die wird aktuell als `device_class=BinarySensorDeviceClass.SAFETY,` angelegt.

Das Teil ist ja grundsätzlich erstmal nur ein Kontakt der OFFFEN/GESCHLOSSEN sein kann. 
So sollte die `HmIP-SCI` meines Erachtens auch per default angelegt werden. 
`BinarySensorDeviceClass.OPENING` würde hier ganz gut passen, denke ich.

Wenn der User dann einen speziellen Anwendungsfall hat, kann er per "Customize" immer noch z.B. einen Fensterkontakt oder Sicherheitskontakt draus machen.

Dann müsste man noch überlegen ob alle HM/IP Taster/Schalterschnittstellen gleichzusetzen wären. Würde ich befürworten und könnte auch wieder alles raussuchen. Aber mir fehlen die echten Geräte um das "Live" zu prüfen.

**Reply by SukramJ** [reply]
Ich denke das kann man machen.

> Dann müsste man noch überlegen ob alle HM/IP Taster/Schalterschnittstellen gleichzusetzen wären. Würde ich befürworten und könnte auch wieder alles raussuchen. Aber mir fehlen die echten Geräte um das "Live" zu prüfen.

Sowas wie ein HmIP-FCI1, HmIP-FCI6? Mach einen Vorschlag, denn die sind bisher noch nicht definiert.

Wenn du mehr Geräte zum testen/anzeigen haben willst, gibt es das Projekt [pydevccu](https://github.com/danielperna84/pydevccu). Das ist eine "Python CCU", die ich auch zum Testen einsetze. Anzeigen geht, aber die Interaktion ist limitiert. Pydevccu kann man starten und in HA einbinden.

**Reply by SukramJ** [reply]
Aktuell sehen die undefinierten binary_sensor so aus:
![image](https://user-images.githubusercontent.com/5650106/152216346-7b88d004-0818-47e4-91af-d591b7174005.png)

**Reply by Baxxy13** [reply]
An/Aus finde ich unpassend. Kommt aber auf den Anwendungsfall an.
(GIF)
![GIF 02 02 2022 19-39-33](https://user-images.githubusercontent.com/54898249/152217262-57ecb75e-88f5-4b68-9ceb-5397dea2d3d0.gif)

**Reply by SukramJ** [reply]
Wenn man vom sprachlichen ausgeht reden wir über Kontaktschnittstellen (Produktbezeichnung).
Also kann man die device_class auf opening setzen also auf  offen / geschlossen.

**Reply by Baxxy13** [reply]
Klingt gut.
Müsste man bei den FCIx dann mal gucken wie das aussieht wenn man das Verhalten der Schnittstelle ändert (Taster/Schalter/Kontakt).

**Reply by SukramJ** [reply]
Gute Frage. Ich habe sowas nicht. 
So ein MULTI_MODE_INPUT_TRANSMITTER hat  PRESS_LONG, PRESS_SHORT und STATE.
Nur STATE wird angezeigt. PRESS_LONG und PRESS_SHORT werden über Device-Trigger abgebildet.
Ob das eine oder andere dann nicht funktioniert?

**Reply by Baxxy13** [reply]
Hmm, ich hab da auch nichts zum testen. 
`MULTI_MODE_INPUT_TRANSMITTER` haben ja quasi alle Geräte mit Eingängen  (z.B. `DRSIx`, `FSI16`). 
Ich würde vermuten das kein Event bei STATE kommt wenn der Eingang auf "Taster" steht.
Spielt aber für die default-Integration eigentlich keine Rolle.
